File: src/types/specialAbility/PactGift.ts
import * as DB from "tsondb/schema/dsl"
import { effect, levels, maximum, name, name_in_library } from "../_Activatable.js"
import { ap_value, ap_value_append, ap_value_l10n } from "../_ActivatableAdventurePointsValue.js"
import { automatic_entries } from "../_ActivatableAutomatic.js"
import { explicit_select_options, select_options } from "../_ActivatableSelectOptions.js"
import { GeneralPrerequisites } from "../_Prerequisite.js"
import { NestedTranslationMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { src } from "../source/_PublicationRef.js"

export const PactGift = DB.Entity(import.meta.url, {
  name: "PactGift",
  namePlural: "PactGifts",
  type: () =>
    DB.Object({
      levels,
      select_options,
      explicit_select_options,
      maximum,
      permanent_demonic_consumption: DB.Optional({
        comment: "This pact gift gives permanent levels of the condition *Demonic Consumption*.",
        type: DB.IncludeIdentifier(PactGiftPermanentDemonicConsumption),
      }),
      automatic_entries,
      prerequisites: DB.Optional({
        type: DB.IncludeIdentifier(GeneralPrerequisites),
      }),
      ap_value,
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "PactGift",
        DB.Object({
          name,
          name_in_library,
          effect,
          ap_value_append,
          ap_value: ap_value_l10n,
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name_in_library",
      keyPathInEntityMapFallback: "name",
    },
  ],
})

const PactGiftPermanentDemonicConsumption = DB.Enum(import.meta.url, {
  name: "PactGiftPermanentDemonicConsumption",
  values: () => ({
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(FixedPactGiftPermanentDemonicConsumption) }),
    PerLevel: DB.EnumCase({
      type: DB.IncludeIdentifier(PactGiftPermanentDemonicConsumptionPerLevel),
    }),
  }),
})

const FixedPactGiftPermanentDemonicConsumption = DB.TypeAlias(import.meta.url, {
  name: "FixedPactGiftPermanentDemonicConsumption",
  type: () =>
    DB.Object({
      levels: DB.Required({
        comment: "The levels of *Demonic Consumption* the pact gift causes.",
        type: DB.Integer({ minimum: 1, maximum: 4 }),
      }),
    }),
})

const PactGiftPermanentDemonicConsumptionPerLevel = DB.TypeAlias(import.meta.url, {
  name: "PactGiftPermanentDemonicConsumptionPerLevel",
  type: () =>
    DB.Object({
      levels: DB.Required({
        comment:
          "The levels of *Demonic Consumption* the pact gift causes per activated level of the pact gift.",
        type: DB.Integer({ minimum: 1, maximum: 4 }),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/specialAbility/PactGift.swift
import FileDB

@Model
public struct PactGift {
      levels,
      select_options,
      maximum,
  /// This pact gift gives permanent levels of the condition *Demonic Consumption*.
  @Relationship(PactGiftPermanentDemonicConsumption)
  let permanent_demonic_consumption: PactGiftPermanentDemonicConsumption.ID?
  /// This pact gift has direct influence on the existence of other entries. It may add or remove entries.
  let automatic_entries: Array(IncludeIdentifier(AutomaticEntry), { minItems: 1 })?
      prerequisites: Optional({
        type: IncludeIdentifier(GeneralPrerequisites),
      }),
      ap_value,

    /// The publications where you can find the entry.
    let src: PublicationRefs

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship
    let translations: [String: Translation]

    struct Translation { // PactGiftTranslation
          name,
          name_in_library,
          rules,

        let errata: Errata?
    }
}

@ModelEnum
public enum PactGiftPermanentDemonicConsumption {
    case Fixed(IncludeIdentifier(FixedPactGiftPermanentDemonicConsumption))
    case PerLevel(IncludeIdentifier(PactGiftPermanentDemonicConsumptionPerLevel))
}

@Embedded
public struct FixedPactGiftPermanentDemonicConsumption {

  /// The levels of *Demonic Consumption* the pact gift causes.
  let levels: Integer({ minimum: 1, maximum: 4 })
  }

@Embedded
public struct PactGiftPermanentDemonicConsumptionPerLevel {

  /// The levels of *Demonic Consumption* the pact gift causes per activated level of the pact gift.
  let levels: Integer({ minimum: 1, maximum: 4 })
  }

@Embedded
public struct AutomaticEntry {

  /// What type of action is applied to the target entry?
  @Relationship(AutomaticEntryAction)
  let action: AutomaticEntryAction.ID

  /// If an entry is added or removed, does is cost or grant adventure points or is it free of charge?
  let apply_ap_value: Boolean()

  /// The entry that is to be added or removed. It can be a fixed entry or a selection where the player must choose one entry.
  @Relationship(AutomaticEntryTarget)
  let target: AutomaticEntryTarget.ID
  }

@ModelEnum
public enum AutomaticEntryAction {
    case Add
    case Remove
}

@ModelEnum
public enum AutomaticEntryTarget {
    case Selection(IncludeIdentifier(AutomaticEntryTargetSelection))
    case Fixed(IncludeIdentifier(FixedAutomaticEntryTarget))
}

@Embedded
public struct AutomaticEntryTargetSelection {
      list: Required({
        type: IncludeIdentifier(AutomaticEntryTargetSelectionList),
      }),
  }

@ModelEnum
public enum AutomaticEntryTargetSelectionList {
    case MagicalTraditions
    case MagicalDilettanteTraditions
}

@Embedded
public struct FixedAutomaticEntryTarget {
      is: Required({
        type: IncludeIdentifier(ActivatableIdentifier),
      }),
  }
